### PR TITLE
refactor: simplify the `make_udf_function` macro

### PR DESCRIPTION
**Pull request by jonahgao**
## Which issue does this PR close?

Follow-up of https://github.com/apache/datafusion/pull/13674#discussion_r1872682452

## Rationale for this change
By limiting the `ScalarUDF` singleton to the function scope, we can remove `$GNAME` from the `make_udf_function` macro. Macro callers no longer need to provide a unique singleton name.

Before:
```rust
make_udf_function!(random::RandomFunc, RANDOM, random);
```
After:
```rust
make_udf_function!(random::RandomFunc, random);
```

## Are these changes tested?
Yes.

## Are there any user-facing changes?
No. Those macros are not exported and are not public APIs.

### CODE REVIEW

File: datafusion/functions/src/macros.rs [hunk 1]
```diff
@@ -65,24 +65,23 @@ macro_rules! export_functions {
...
-/// function named `$NAME` which returns that singleton.
+/// Creates a singleton `ScalarUDF` of the `$UDF` function and a function
+/// named `$NAME` which returns that singleton.
 ///
 /// This is used to ensure creating the list of `ScalarUDF` only happens once.
 macro_rules! make_udf_function {
-    ($UDF:ty, $GNAME:ident, $NAME:ident) => {
-        #[doc = "Return a [`ScalarUDF`](datafusion_expr::ScalarUDF) implementation "]
-        #[doc = stringify!($UDF)]
+    ($UDF:ty, $NAME:ident) => {
```
Comment (jonahgao) [inline datafusion/functions/src/macros.rs:73]: Here is the main change.

File: datafusion/functions/src/macros.rs [hunk 2]
```diff
@@ -65,24 +65,23 @@ macro_rules! export_functions {
...
+        #[doc = concat!("Return a [`ScalarUDF`](datafusion_expr::ScalarUDF) implementation of ", stringify!($NAME))]
         pub fn $NAME() -> std::sync::Arc<datafusion_expr::ScalarUDF> {
             // Singleton instance of the function
-            static $GNAME: std::sync::LazyLock<
+            static INSTANCE: std::sync::LazyLock<
```
Comment (comphead) [inline datafusion/functions/src/macros.rs:77]: here we go where lazy lock shines

**Review by comphead — APPROVED** [review]
lgtm thanks @jonahgao

**Review by jayzhan211 — APPROVED** [review]

**Comment by jayzhan211** [comment]
Thanks @jonahgao @comphead

**Comment by jonahgao** [comment]
Thanks @comphead @jayzhan211 for the review

**Pull request merged**